src/descriptor/satisfied_contraints.rs

```diff
@@ -0,0 +1,1610 @@
+/// All stack elements with vec![] go to Dissatisfied and vec![1] are marked to Satisfied.
+/// Others are directly pushed as witness
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub enum StackElement<'stack> {
+    /// Result of a satisfied E/T/W/F
+    Satisfied,
+    /// Result of a dissatisfied E/W
+    Dissatisfied,
+    /// Input from the witness stack
+    Push(&'stack [u8]),
+}
+
+impl<'stack> StackElement<'stack>{
+
+    pub fn from(v: &'stack [u8]) -> StackElement<'stack> {
+        if *v == [1]{
+            StackElement::Satisfied
+        } else if *v == [] {
+            StackElement::Dissatisfied
+        } else {
+            StackElement::Push(v)
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum HashLockType<'desc>{
+    ///SHA 256 hashlock
+    Sha256(&'desc sha256::Hash),
+    ///Hash 256 hashlock
+    Hash256(&'desc sha256d::Hash),
+    ///Hash160 hashlock
+    Hash160(&'desc hash160::Hash),
+    ///Ripemd160 hashlock
+    Ripemd160(&'desc ripemd160::Hash),
+}
+/// A satisfied Miniscript condition (Signature, Hashlock, Timelock)
+/// 'desc represents the lifetime of descriptor and `stack represents the lifetime of witness
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum SatisfiedConstraint<'desc, 'stack, Pk: 'desc, Pkh: 'desc> {
+    ///Public key and corresponding signature
+    PublicKey {
+        key: &'desc Pk,
+        sig: secp256k1::Signature,
+    },
+    ///PublicKeyHash, corresponding pubkey and signature
+    PublicKeyHasH {
+        keyhash: &'desc Pkh,
+        key: bitcoin::PublicKey,
+        sig: secp256k1::Signature,
+    },
+    ///Hashlock and preimage for SHA256
+    HashLock {
+        hash: HashLockType<'desc>,
+        preimage: &'stack [u8],
+    },
+    ///Relative Timelock for CSV.
+    RelativeTimeLock {
+        time: &'desc u32,
+    },
+    ///Absolute Timelock for CLTV.
+    AbsoluteTimeLock {
+        time: &'desc u32,
+    },
+}
+
+///This is used by the interpreter to know which evaluation state a AstemElem is.
+///This is required because whenever a same node(for eg. OrB) appears on the stack, we don't
+///know if the left child has been evaluated or not. And based on the result on
+///the top of the stack, we need to decide whether to execute right child or not.
+///This is also useful for wrappers and thresholds which push a value on the stack
+///depending on evaluation of the children.
+struct NodeEvaluationState<'desc, Pk: 'desc, Pkh: 'desc> {
+    ///The node which is being evaluated
+    node: &'desc AstElem<Pk, Pkh>,
+    ///number of children evaluated
+    n_evaluated: usize,
+    ///number of children satisfied
+    n_satisfied: usize,
+}
+
+/// An iterator over all the constraints satisfied by a given
+/// descriptor/scriptSig/witness stack tuple
+pub struct SatisfiedConstraints<'desc, 'stack, Pk: 'desc, Pkh: 'desc> {
+    secp: &'desc secp256k1::Secp256k1<VerifyOnly>,
```

Comment:
I know we have a lot of lifetimes already but we should probably give `secp` its own